### PR TITLE
scx_layered: Fix crash on aarch64 due to unavailable cache id file

### PR DESCRIPTION
**Pull request by minosfuture**
This PR addresses #733

* scx_layered: Read CPU topology for building CpuPool
  * Building CpuPool from cache-cpu topology did not apply on arm, because `/sys/devices/system/cpu/cpu{}/cache/index{}/id` file is unavailable.
  * Read CPU topology from topology.rs instead.

### CODE REVIEW

**Comment by htejun** [comment]
Thanks for working on this but would it be possible to move all the topology logic into `scx_utils::topology`?

**Review by hodgesds — COMMENTED** [review]
Looks good overall, if you move some of the helper functions to the topology crate that would be useful elsewhere. `CpuPool` is layered specific so anything touching that can probably stay.

**Review by JakeHillion — COMMENTED** [review]
Agreed with @htejun here, this sort of read-once (modulo hotplugs) stuff would be nicer in topology, and topology already has the right mask for Core -> Cpus (https://github.com/sched-ext/scx/blob/4979cb87629888734f287f55cdf63f0072cb1b94/rust/scx_utils/src/topology.rs#L188-L191) - though I can't see where it gets set...

As an aside, if you throw the comments commit in a separate PR that can likely be merged - it's desperately needed and thanks for doing it!

**Comment by minosfuture** [comment]
Sounds good. I'll rework this. Let's get the doc in first: #754.

**Comment by minosfuture** [comment]
@htejun @hodgesds @JakeHillion could you take another look? Util functions are now moved to topology.rs.
Thanks!

File: rust/scx_utils/src/topology.rs [hunk 1]
```diff
@@ -416,6 +461,17 @@ impl TopologyMap {
     pub fn iter(&self) -> Iter<Vec<usize>> {
         self.map.iter()
     }
+
+    pub fn core_cpus_bitvec(&self) -> Vec<BitVec> {
```
Comment (hodgesds) [inline rust/scx_utils/src/topology.rs:465]: nit: comment what this returns

**Review by hodgesds — APPROVED** [review]
Seems good overall, thanks for cleaning this up!

**Comment by minosfuture** [comment]
Could you share the testing method? Thanks.

**Comment by minosfuture** [comment]
> @minosfuture if you have access to a multi NUMA machine can you try running the [integration test](https://github.com/sched-ext/scx/blob/main/scheds/rust/scx_layered/integration/run_tests.sh) I've added for `scx_layered`?

I don't have one unfortunately. Can I reproduce a version of the problem of cpumask you were mentioning in some other way? @hodgesds

**Comment by hodgesds** [comment]
@minosfuture if you rebase this diff so it builds again I'll do some more testing on NUMA machines. Hopefully we'll have some NUMA machines in the CI pipeline soon.

**Comment by minosfuture** [comment]
@hodgesds rebased and updated.
thanks again for testing this!
to reproduce the cpumask issue, is numa required?

**Comment by hodgesds** [comment]
I still see the integration tests failing with your rebase and it reproduces every time.
```
###### Tue, 15 Oct 2024 08:55:34 -0700 ######
tot=  53526 local=89.51 open_idle= 0.06 affn_viol= 0.00 proc=5ms nodes=2
busy= 12.7 util=  987.7 load=     15.2 fallback_cpu=  0
excl_coll=0.00 excl_preempt=0.00 excl_idle=0.00 excl_wakeup=0.00
  numa_0: util/dcycle/frac/302.0/ 30.6/   30.6 load/load_frac_adj/frac=     3.05/30.58/ 20.0 tasks=    57
          tot=   6247 local=49.99 wake/exp/reenq= 0.00/50.01/ 0.00
          xlayer_wake= 0.05 xlayer_rewake= 0.00
          keep/max/busy= 1.46/ 0.06/ 0.00 kick=50.01 yield/ign= 0.06/    0 
          open_idle= 0.00 mig= 2.61 xnuma_mig= 0.80 xllc_mig= 0.80 affn_viol= 0.00
          preempt/first/xllc/xnuma/idle/fail= 0.00/ 0.00/ 0.00/ 0.00/ 0.00/ 0.00 min_exec= 0.00/   0.00ms, slice=20ms
          cpus=  0 [  0,  0] 00000000 00000000 00000000
  reset : util/dcycle/frac/685.7/ 69.4/   69.4 load/load_frac_adj/frac=    12.17/69.42/ 80.0 tasks=  3188
          tot=  47279 local=94.73 wake/exp/reenq= 4.92/ 0.35/ 0.00
          xlayer_wake=26.78 xlayer_rewake= 6.65
          keep/max/busy= 0.45/ 0.01/ 0.00 kick= 0.35 yield/ign= 0.15/    0 
          open_idle= 0.07 mig=10.03 xnuma_mig= 2.54 xllc_mig= 2.54 affn_viol= 0.00
          preempt/first/xllc/xnuma/idle/fail= 0.00/ 0.00/ 0.00/ 0.00/ 0.00/ 0.00 min_exec= 0.00/   0.00ms, slice=20ms
          cpus=  0 [  0,  0] 00000000 00000000 00000000
./run_tests.sh: line 7: 2483700 Killed                  sudo "./${test_script}" "${stress_pid}"
2480015
./run_tests.sh: line 7: 2480002 Killed                  sudo "${layered_bin}" --stats 1 "f:${test_config}" -v
./run_tests.sh: line 7: 2490412 Killed                  sudo pkill -9 -f scx_layered
./run_tests.sh: line 7: 2490589 Killed                  sudo pkill -9 -f stress-ng
test script layer_node.bt failed: 137
```
cc @JakeHillion to take a look at the core id mapping, which I think is the issue.

**Comment by hodgesds** [comment]
> @hodgesds rebased and updated. thanks again for testing this! to reproduce the cpumask issue, is numa required?

I think if you have a machine with multiple LLCs you can run the integration test as well. Here's some general steps for how to run it:
```
cd scx
cargo build --release
cd scheds/rust/scx_layered/integration
./run_tests.sh
```

**Comment by minosfuture** [comment]
@hodgesds I found a issue in generating `cpu_core`. I had a wrong understanding that I thought it was the physical core id list, but it's actually the physical core id for each logical core (of the index). This is fixed, and the `CpuPool` values are identical on my amd64 test machine.

I'm having the following issue so I still could running the integration test.  I'll investigate further.  In the meantime, do you mind testing this again and see if we can get it in? Thank you.
```bash
./layer_llc.bt:31:17-31: ERROR: Struct/union of type 'struct cacheinfo' does not contain a field named 'id'
        $l3_cacheid = $l3_cacheinfo.id;
                      ~~~~~~~~~~~~~~
98171
test script layer_llc.bt failed: 1
```

**Review by hodgesds — APPROVED** [review]
I reran the integration tests on a multi NUMA machine and everything passes. Thanks for making this work!

```
$  ./run_tests.sh 
...
###### Sat, 19 Oct 2024 10:05:34 -0700 ######
tot=  38425 local=89.30 open_idle=27.51 affn_viol= 0.00 proc=5ms nodes=2
busy=  6.6 util=  506.1 load=      9.1 fallback_cpu=  3
excl_coll=0.00 excl_preempt=0.00 excl_idle=0.00 excl_wakeup=0.00
  llc_0: util/dcycle/frac/304.4/ 60.1/   60.1 load/load_frac_adj/frac=     3.05/60.14/ 33.5 tasks=   168
         tot=   5983 local=49.61 wake/exp/reenq= 0.33/50.06/ 0.00
         xlayer_wake= 0.03 xlayer_rewake= 0.00
         keep/max/busy= 1.47/ 0.05/ 0.08 kick=49.41 yield/ign= 0.00/    0 
         open_idle= 0.00 mig= 0.80 xnuma_mig= 0.00 xllc_mig= 0.00 affn_viol= 0.00
         preempt/first/xllc/xnuma/idle/fail= 0.00/ 0.00/ 0.00/ 0.00/ 0.00/ 0.00 min_exec= 0.00/   0.00ms, slice=20ms
         cpus=  4 [  4,  4] 00000003 00000300 00000000
  reset: util/dcycle/frac/201.8/ 39.9/   39.9 load/load_frac_adj/frac=     6.05/39.86/ 66.5 tasks=  3669
         tot=  32442 local=96.62 wake/exp/reenq= 3.21/ 0.16/ 0.00
         xlayer_wake=37.16 xlayer_rewake= 9.29
         keep/max/busy= 0.15/ 0.00/ 0.00 kick= 0.16 yield/ign= 0.09/    0 
         open_idle=32.58 mig=16.53 xnuma_mig= 3.92 xllc_mig= 3.92 affn_viol= 0.00
         preempt/first/xllc/xnuma/idle/fail= 0.00/ 0.00/ 0.00/ 0.00/ 0.00/ 0.00 min_exec= 0.00/   0.00ms, slice=20ms
         cpus= 10 [ 10, 10] 00000004 000004f0 0000f000
2040932
./run_tests.sh: line 7: 2040930 Killed                  sudo "${layered_bin}" --stats 1 "f:${test_config}" -v
./run_tests.sh: line 7: 2102275 Killed                  sudo pkill -9 -f scx_layered
./run_tests.sh: line 7: 2102914 Killed                  sudo pkill -9 -f stress-ng
test script layer_llc.bt passed: 0
```

**Pull request merged**